realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -66,11 +94,16 @@ public void executeDelayedDisposal() {
         }
     }
 
-    public void cleanNativeReferences() {
+    private void cleanNativeReferences() {
         NativeObjectReference reference = (NativeObjectReference) referenceQueue.poll();
         while (reference != null) {
+            // Dealloc the native resources
+            reference.cleanup();
             reference.clear();
-            rowReferences.remove(reference);
+            // Inline referencesPool.freeIndexList.remove() to make it faster.
+            // referencesPool.freeIndexList.arrayList.set(index, null); is not really needed. Make it faster by not
+            // setting the slot to null.
+            referencesPool.freeIndexList.add(reference.refIndex);
```

Comment:
Yeah, i tried to pack them in one method, but according to the test result, it slows down thing ~10% on Davik .... Java is kind of unpredictable language ...